doc/horizontal_autoscaling.md

```diff
@@ -0,0 +1,163 @@
+# Design Doc: Horizontal Autoscaling
+
+## Background
+
+A PaddlePaddle training job contains several trainer instances,
+several parameter server instances, and one master instance. We would
+like to automatically scale the number of training instances and the
+number of parameter server instances to fully utilize the cluster's
+computation resources.
+
+Currently, we will only support horizontal trainer
+autoscaling. Parameter server horizontal autoscaling will be supported
+in the near future. This design doc considers both of them.
+
+[Horizontal Pod Autoscaling (HPA)](https://kubernetes.io/docs/tasks/run-application/horizontal-pod-autoscale/) is
+an autoscaling solution provided by Kuberentes, but it's not suitable
+for the training job autoscaling for the following reasons:
+
+- The training job autoscaling tries to fairly distribute the
+  computation resources to different training jobs, in order to
+  optimize the computation resource utilization of the **entire**
+  cluster. The HPA is trying to improve the quality of service of a
+  **single** service. The training job autoscaling requires the
+  controller to have a global view of all the available computation
+  resources and all the training jobs, but HPA does not have the
+  global view.
+
+- HPA is designed to automatically scale a homogeneous set of Pods,
+  but we need to scale a heterogeneous set of Pods (the trainer Pods
+  and the parameter server Pods): because the required number of
+  parameter servers is correlated to the required number of trainers,
+  we need to scale them together.
+
+We need to develop our own solution for autoscaling.
+
+## Solution
+
+We will build
+an [Operator](https://coreos.com/blog/introducing-operators.html) that
+do the autoscaling. To be more precise, we will create
+a
+[custom Kubernetes Resource](https://kubernetes.io/docs/concepts/api-extension/custom-resources/) and
+a
+[custom Kuberentes Controller](https://resources.coreos.com/youtube-coreos-fest-2017/writing-a-custom-controller-extending-the-functionality-of-your-cluster). Kuberentes
+is designed to be flexible so adding custom Resources and custom
+Controllers into our cluster does not require modifying the Kubernetes
+source code.
+
+### Training Job Resource
+
+Just
+like
+[Deployment](https://kubernetes.io/docs/concepts/workloads/controllers/deployment/) is
+a resource that describes a deployment. We will have a training
+job
+[Custom Resource](https://kubernetes.io/docs/concepts/api-extension/custom-resources/) that
+describes the training job.
+
+A pseudo resource declaration (`training_job.yaml`) is as follows:
+
+```yaml
+apiVersion: cloud.paddlepaddle.org/v1
+kind: TrainingJob
+metadata:
+  name: job-1
+spec:
+  trainer:
+    entrypoint: "python train.py"
+    workspace: "/home/job-1/"
+    requests: 3
+    limits: 6
+    cpu: "500m"
```

Comment:
you can embed k8s job spec if you want.

Reply:
Thanks for the heads up! Maybe we can make the description independent of the Job spec, so we are not obliged to run the trainers with a Job.